src/views/Predictions/components/RoundCard/SetPositionCard.tsx

```diff
@@ -73,19 +76,38 @@ const getValueAsEthersBn = (value: string) => {
   return Number.isNaN(valueAsFloat) ? Zero : parseUnits(value)
 }
 
+const TOKEN_BALANCE_CONFIG = {
+  BNB: useGetBnbBalance,
+  CAKE: useGetCakeBalance,
+}
+
 const SetPositionCard: React.FC<SetPositionCardProps> = ({ position, togglePosition, epoch, onBack, onSuccess }) => {
   const [value, setValue] = useState('')
   const [errorMessage, setErrorMessage] = useState(null)
   const [percent, setPercent] = useState(0)
 
   const { account } = useWeb3React()
-  const { balance: bnbBalance } = useGetBnbBalance()
   const minBetAmount = useGetMinBetAmount()
   const { t } = useTranslation()
   const { fetchWithCatchTxError, loading: isTxPending } = useCatchTxError()
   const { callWithGasPrice } = useCallWithGasPrice()
   const { address: predictionsAddress, token } = useConfig()
-  const predictionsContract = usePredictionsContract(predictionsAddress)
+  const predictionsContract = usePredictionsContract(predictionsAddress, token.symbol)
+  const useTokenBalance = useMemo(() => {
+    return TOKEN_BALANCE_CONFIG[token.symbol]
+  }, [token.symbol])
+
+  const { isVaultApproved, setLastUpdated } = useCakeApprovalStatus(token.symbol === 'CAKE' ? predictionsAddress : null)
+  const { handleApprove, pendingTx } = useCakeApprove(
+    setLastUpdated,
+    predictionsAddress,
+    t('You can now start prediction'),
+  )
+
+  // BNB prediction doesn't need approval
+  const doesCakeApprovePrediction = token.symbol === 'BNB' || isVaultApproved
```

Comment:
Why don't we use useApproveConfirmTransaction hook?

Reply:
AFAIK, `useApproveConfirmTransaction` is low-level hook which is used for more generic cases. If we use `useApproveConfirmTransaction` for such specific cases, the duplication would increase.